Storage/src/Connection/Rest.php

```diff
@@ -628,6 +629,53 @@ private function sendWithRetry($resource, $method, $args)
             $args
         );
 
+        $retryIdentifierHash = Uuid::uuid4()->toString();
+
+        $args['onRetryException'] = function (\Exception $e, $currentAttempt, &$arguments) use ($retryIdentifierHash) {
+            // Since we the the last attempt number here, so incrementing it
+            // to get the current attempt count
+            ++$currentAttempt;
+            $this->updateRetryInovationHeaders(
+                $arguments,
+                $retryIdentifierHash,
+                $currentAttempt
+            );
+        };
+
+        $args['onExecutionStart'] = function (&$arguments) use ($retryIdentifierHash) {
+            $this->updateRetryInovationHeaders(
+                $arguments,
+                $retryIdentifierHash
+            );
+        };
+
         return $this->send($resource, $method, $args);
     }
+
+    /**
+     * Updates the x-goog-api-client header value with UUID
+     * and retry count
+     *
+     * @return void
+     */
+    private function updateRetryInovationHeaders(
+        &$arguments,
+        $retryIdentifierHash,
+        $currentAttempt = 0
+    ) {
+        // Fetch existing value(if present) of ['x-goog-api-client']
+        // header value is present and create new required value
+        $request = $arguments[0];
+        $headerValue = $request->getHeaderLine('x-goog-api-client');
+        $headerValue = $headerValue ? $headerValue . " " : "";
+        $headerValue = $headerValue .
+            "gccl-invocation-id/$retryIdentifierHash " .
+            "gccl-attempt-count/$currentAttempt";
```

Comment:
Probably cleaner:

```suggestion
        $headers = [
            $headerValue,
            sprintf('gccl-invocation-id/%s',$retryIdentifierHash),
            sprintf('gccl-attempt-count/%s',$currentAttempt)
         ];
        $headerValue = implode(' ', $headerValue);
```

Reply:
Yes, I used the same here [af0e43c](https://github.com/googleapis/google-cloud-php/pull/5741/commits/af0e43c40fb3d06f1eb1649a437cf04305ab765a)